### PR TITLE
Add back writeTo(Filer) and originating element APIs

### LINKED ISSUE #105: Consider adding Filer convenience methods back
kapt compiles kt files generated from processors per comments on https://github.com/square/kotlinpoet/pull/93

### PR DESCRIPTION
**Pull request by ZacSweers**
Resolves #105

This adds back the `FileSpec#writeTo(File)` API and originating element APIs on `TypeSpec`, as well as spreading support for originating elements onto `PropertySpec` and `FunSpec` (since you could have files with no types). Most of this was borrowed from the original revert in  d24e4afb01cd9e933f40835cee2a3ffe90a598f5

This uses a simpler version of the interface pattern `Taggable` uses to reuse the logic and make integrating it simple in `FileSpec`.

### CODE REVIEW

**Comment by ZacSweers** [comment]
Will rebase this tomorrow

File: src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    /** Adds an [originatingElement] to this type's list of originating elements. */
+    @Suppress("UNCHECKED_CAST")
+    fun addOriginatingElement(originatingElement: Element): T = apply {
+      originatingElements += originatingElement
+    } as T
+  }
+}
+
+internal fun OriginatingElementsHolder.Builder<*>.buildOriginatingElements() = OriginatingElements(originatingElements.toImmutableList())
```
Comment (Egorand) [inline src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt]: nit: newline after `=`?

File: src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+internal fun List<Element>.buildOriginatingElements() = OriginatingElements(toImmutableList())
+
+internal class OriginatingElements(override val originatingElements: List<Element>) : OriginatingElementsHolder
```
Comment (Egorand) [inline src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt]: nit: newline after `=`?

File: src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt [hunk 3]
```diff
@@ -2421,11 +2422,15 @@ class TypeSpecTest {
...
+        .addOriginatingElement(originatingElement)
         .build()
 
-    assertThat(comprehensiveTaco.toBuilder().build()).isEqualTo(comprehensiveTaco)
+    val newTaco = comprehensiveTaco.toBuilder().build()
+    assertThat(newTaco).isEqualTo(comprehensiveTaco)
+    assertThat(newTaco.originatingElements).containsExactly(originatingElement)
   }
 
+
```
Comment (Egorand) [inline src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt]: nit: empty line

File: src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package com.squareup.kotlinpoet
```
[the remark this answers is not in the document]
Comment (Egorand) [reply src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt]: It does, good catch!

File: src/test/java/com/squareup/kotlinpoet/TestFiler.kt [hunk 5]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2014 Square, Inc.
```
[the remark this answers is not in the document]
Comment (Egorand) [reply src/test/java/com/squareup/kotlinpoet/TestFiler.kt]: Good point, although it's a revert of a previously deleted file, it'd make sense to use 2019

File: src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt (hunk 2 above)
Comment (ZacSweers) [reply src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt]: you mean after `:`?

File: src/test/java/com/squareup/kotlinpoet/TestFiler.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (JakeWharton) [reply src/test/java/com/squareup/kotlinpoet/TestFiler.kt]: I would keep 2014. This is like copying a file from another project. Leave it as is unless it's actually been _written_ by hand.

[the remark this answers is not in the document]
Comment (ZacSweers) [reply src/test/java/com/squareup/kotlinpoet/TestFiler.kt]: Part of 38960b5

File: src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (ZacSweers) [reply src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt]: Part of 38960b5

File: src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt (hunk 3 above)
Comment (ZacSweers) [reply src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt]: Part of 38960b5

File: src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt (hunk 2 above)
Comment (ZacSweers) [reply src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt]: Part of 38960b5

File: src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt (hunk 1 above)
Comment (ZacSweers) [reply src/main/java/com/squareup/kotlinpoet/OriginatingElementsHolder.kt]: Part of 38960b5

File: src/main/java/com/squareup/kotlinpoet/Taggable.kt [hunk 6]
```diff
@@ -1,3 +1,18 @@
+/*
```
Comment (Egorand) [inline src/main/java/com/squareup/kotlinpoet/Taggable.kt]: I'd prefer this to be in a separate PR

File: src/test/java/com/squareup/kotlinpoet/TestFiler.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (ZacSweers) [reply src/test/java/com/squareup/kotlinpoet/TestFiler.kt]: Whoops didn't see Jake's comment before pushing. Back to 2014 in da3d8ef

File: src/main/java/com/squareup/kotlinpoet/Taggable.kt (hunk 6 above)
Comment (ZacSweers) [reply src/main/java/com/squareup/kotlinpoet/Taggable.kt]: ea806fc
